### PR TITLE
Improve PyPy compatibility

### LINKED ISSUE #4701: Improve PyPy compatibility
Running the following code using PyPy and Jupyter notebook will raise several exceptions:

```
import numpy as np
import random
from bokeh.plotting import figure, show
from bokeh.io import output_notebook
import bokeh

# plot inline
output_notebook()

# prepare some data
x = range(100)
y = [random.randint(0,100) for i in x]

# create a new plot with a title and axis labels
p = figure(title="simple line example", x_axis_label='x', y_axis_label='y')

# add a line renderer with legend and line thickness
p.line(x, y, legend="Temp.", line_width=2)

# show the results
show(p)
```

The issue is related to the fact that numpy for PyPy doesn't have np.datetime64 implemented yet.
The solution is quite easy to implement and it is roughly described here: [http://stackoverflow.com/questions/37570376/is-there-any-2d-plotting-library-compatible-with-pypy/37588552#37588552](http://stackoverflow.com/questions/37570376/is-there-any-2d-plotting-library-compatible-with-pypy/37588552#37588552)

### PR DESCRIPTION
**Pull request by alec-djinn**
- [ ] issues: fixes #4701

### CODE REVIEW

**Comment by bryevdv** [comment]
We have installed a fairly strict linter, there's a line with trailing whitespace that is failing the tests:

```
E        +  where 1 = len(['File contains trailing whitespace: bokeh/core/properties.py, line 1555.'])
```

**Comment by bryevdv** [comment]
Also an indentation error reported:
```
  File "/home/travis/miniconda/lib/python2.7/site-packages/bokeh/util/serialization.py", line 68
    if 'PyPy' in sys.version:
```

File: bokeh/util/serialization.py [hunk 1]
```diff
@@ -57,9 +57,22 @@ def transform_array(obj):
...
-                         dt2001.astype('datetime64[ms]').astype('int64'))
+    try:
+        dt2001 = np.datetime64('2001')
+        legacy_datetime64 = (dt2001.astype('int64') ==
+                             dt2001.astype('datetime64[ms]').astype('int64'))
+    ## For compatibility with PyPy that doesn't have datetime64
+    except AttributeError as e:
+        if e.args == ("'module' object has no attribute 'datetime64'",):
+                import sys
+                 if 'PyPy' in sys.version:
```
Comment (bryevdv) [inline bokeh/util/serialization.py]: Here is the indentation error, please also indent this block the usual 4 spaces

**Comment by bryevdv** [comment]
BTW do you know whether the unit test suite otherwise passes under PyPy? I would not be opposed to added a PyPy unit test run to the CI infrastructure if so.

**Comment by bryevdv** [comment]
Also, thanks for the PR :)

File: bokeh/util/serialization.py (hunk 1 above)
Comment (alec-djinn) [inline bokeh/util/serialization.py]: sorry, my bad!

**Comment by alec-djinn** [comment]
>> do you know whether the unit test suite otherwise passes under PyPy?
I have no idea...

**Comment by bryevdv** [comment]
I don't have `PyPy` installed but if you do, you should be able to check by just running
```
import bokeh; bokeh.test()
```

**Comment by alec-djinn** [comment]
```
[PyPy 5.1.1 with GCC 4.2.1 Compatible Apple LLVM 7.3.0 (clang-703.0.31)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>>> import bokeh
>>>> bokeh.test()
============================= test session starts ==============================
platform darwin -- Python 2.7.10[pypy-5.1.1-final], pytest-2.9.2, py-1.4.31, pluggy-0.3.1
rootdir: /Users/.../homebrew/lib/pypy/site-packages/bokeh, inifile: 
collected 0 items / 1 errors 

==================================== ERRORS ====================================
______________________________ ERROR collecting  _______________________________
../../../../../Homebrew/Cellar/pypy/5.1.1_1/libexec/site-packages/py/_path/common.py:332: in visit
    for x in Visitor(fil, rec, ignore, bf, sort).gen(self):
../../../../../Homebrew/Cellar/pypy/5.1.1_1/libexec/site-packages/py/_path/common.py:368: in gen
    if p.check(dir=1) and (rec is None or rec(p))])
../../../../../Homebrew/Cellar/pypy/5.1.1_1/libexec/site-packages/_pytest/main.py:647: in _recurse
    ihook = self.gethookproxy(path)
../../../../../Homebrew/Cellar/pypy/5.1.1_1/libexec/site-packages/_pytest/main.py:550: in gethookproxy
    my_conftestmodules = pm._getconftestmodules(fspath)
../../../../../Homebrew/Cellar/pypy/5.1.1_1/libexec/site-packages/_pytest/config.py:297: in _getconftestmodules
    mod = self._importconftest(conftestpath)
../../../../../Homebrew/Cellar/pypy/5.1.1_1/libexec/site-packages/_pytest/config.py:322: in _importconftest
    raise ConftestImportFailure(conftestpath, sys.exc_info())
E   ConftestImportFailure: (local('/Users/.../homebrew/lib/pypy/site-packages/bokeh/charts/conftest.py'), (<type 'exceptions.RuntimeError'>, RuntimeError('The bokeh.charts interface requires Pandas (http://pandas.pydata.org) to be installed.',), <traceback object at 0x0000000107cad840>))
=========================== 1 error in 0.39 seconds ============================
1
```


Pandas is needed to pass the test... Pandas is still not compatible with PyPy.
But for easy plots, you don't need it... You can consider making a minimal version just for PyPy using only the standard library and numpy.

**Comment by bryevdv** [comment]
Pandas is optional and only needed for `bokeh.charts`, we could make those tests conditional.

**Comment by alec-djinn** [comment]
It worths a try...
Also, np.datime64 is not yet implemented in the pypy version of numpy... that will also give problems.
In my case, I have solved with the two PR I am working on now, but I am not sure if other cases (more fancy plots) will be covered as well.

**Comment by bryevdv** [comment]
OK, the last failing test is expected (sauce labs tests can only run for @bokeh/dev PRs). I'd be OK merging this for now, if you are content with it. In that case can you provide a screen shot of a PyPy created plot here just for the record? I'm also happy wait. if you'd like to work on this PR longer to try and get all the unit tests passing. There's no hurry on our end so it's really up to you.

**Comment by alec-djinn** [comment]
Sure! I'll provide the screenshot asap.

**Comment by alec-djinn** [comment]
I have pushed the last modification on properties.py, also timedelta64 gives some problems, not all looks fine to me.
Here the screenshots before and after the updates.

after
before

**Comment by alec-djinn** [comment]
I don't get why #9150.1 failed...

**Comment by bryevdv** [comment]
We have some issues with intermittent spurious test failures. I have restarted that job.

**Comment by bryevdv** [comment]
OK, merging now. Thanks @alec-djinn

**Comment by alec-djinn** [comment]
I am happy to help!

**Comment by mrocklin** [comment]
Thanks for doing this @alec-djinn !  I was looking at whether or not Bokeh supported PyPy and was glad to see your work here:

**Pull request merged**